### ISSUE TITLE
Can we implement a search using async data from a server?

### ISSUE THREAD

**Issue by fayazara** [ssleptsov/ninja-keys#36]
We're planning on making the ninja-keys as the primary search window for our tool. It works well for items which are defined already, but is there a way to make api requests and get data and show them in Ninja keys command palette?

CC: @ssleptsov

**Comment by eelcoj** [comment]
@fayazara Maybe a bit late, but this exactly what I've done for my product. Needs a bit of work, but not too complicated.

<img width="779" alt="search" src="https://user-images.githubusercontent.com/988051/215390931-3a01fd4d-37f3-41a5-8022-a4cc944fbeb7.png">

Anything below "Content Types" is static. Then below that you have all the search results.

**Comment by elalemanyo** [comment]
@eelcoj how did you do it? did you use ninja-keys?

**Comment by eelcoj** [comment]
@elalemanyo Yes, all it really does is fire a request to backend's search. The json response is then changed to match the data structure ninja keys need.

**Comment by elalemanyo** [comment]
@eelcoj can you show me how you are doing it? if it is possible. Thanks

**Comment by eelcoj** [comment]
@elalemanyo Would only be useful if you use the same framework, right? In essence listen for the change event on the input field, when certain character threshold is met (eg. min. 3 characters), send async post request to get the search results, than merge this result into the data structure needed for ninja-keys.

**Comment by dandv** [comment]
> is there a way to make api requests and get data and show them in Ninja keys command palette?

Yes, but it's not exactly designed for that. There are better tools for the job.

Here's how you'd do it with NinjaKeys:

```js
ninja.addEventListener('change', (event) => {
  debouncedSearch(event.detail.search);  // define your debounced search function
})
```

One problem with this is that Ninja filters the `data` array for titles that match the search queries. For keyword searches this is fine, but if you run a <a href="https://jdayusa.dandv.me/">semantic search</a> and your result doesn't echo the query (e.g. search for "excursion" and find "trip") then Ninja won't show the result!

One library that displays exactly a search modal is [search-modal](https://github.com/4msar/search-modal). Needs some love but does the job more cleanly. You just need to supply the search URL.

**Comment by san425** [comment]
> > is there a way to make api requests and get data and show them in Ninja keys command palette?
> 
> Yes, but it's not exactly designed for that. There are better tools for the job.
> 
> Here's how you'd do it with NinjaKeys:
> 
> ```js
> ninja.addEventListener('change', (event) => {
>   debouncedSearch(event.detail.search);  // define your debounced search function
> })
> ```
> 
> One problem with this is that Ninja filters the `data` array for titles that match the search queries. For keyword searches this is fine, but if you run a [semantic search](https://jdayusa.dandv.me/) and your result doesn't echo the query (e.g. search for "excursion" and find "trip") then Ninja won't show the result!
> 
> One library that displays exactly a search modal is [search-modal](https://github.com/4msar/search-modal). Needs some love but does the job more cleanly. You just need to supply the search URL.

Its a great library, but you can easily override the ctrl + K; it opens by default on ctrl + K / cmd + K

**Comment by olivercoad** [comment]
> One problem with this is that Ninja filters the `data` array

Could you get around that by adding the search query as `keyword` in data array for all search results so that it always matches?
